planetiler-custommap/src/main/java/com/onthegomap/planetiler/custommap/ConfiguredFeature.java

```diff
@@ -0,0 +1,302 @@
+      case bool:
+        return booleanTypeConverter();
+      //Default: pass through
+      default:
+        return in -> in;
+    }
+  }
+
+  private static Set<String> booleanTrue = new HashSet<>(Arrays.<String>asList("true", "yes", "1"));
+  private static Set<String> booleanFalse = new HashSet<>(Arrays.<String>asList("false", "no", "0"));
+
+  /**
+   * Convert a raw tag value to a corresponding normalized boolean value.
+   * 
+   * @return normalization function
+   */
+  private static Function<Object, Object> booleanTypeConverter() {
+    return s -> {
+      if (booleanTrue.contains(s)) {
+        return Boolean.TRUE;
+      }
+      if (booleanFalse.contains(s)) {
+        return Boolean.FALSE;
+      }
+      return null;
+    };
+  }
+
+  /**
+   * Generate logic which determines the minimum zoom level for a feature based on a configured pixel size limit.
+   * 
+   * @param minTilePercent - minimum percentage of a tile that a feature must cover to be shown
+   * @param minZoom        - global minimum zoom for this feature
+   * @return minimum zoom function
+   */
+  private static Function<SourceFeature, Integer> attributeZoomThreshold(double minTilePercent, int minZoom) {
+    return sf -> {
+      try {
+        int zoom = (int) (Math.log(minTilePercent / sf.area()) / LOG4);
+        return Math.max(minZoom, Math.min(zoom, 14));
+      } catch (GeometryException e) {
+        return 14;
+      }
+    };
+  };
+
+  /**
+   * Generates a function which produces a fully-configured attribute for a feature.
+   * 
+   * @param attribute - configuration for this attribute
+   * @return processing logic
+   */
+  private BiConsumer<SourceFeature, Feature> attributeProcessor(AttributeDefinition attribute) {
+    String tagKey = attribute.getKey();
+    Integer configuredMinZoom = attribute.getMinZoom();
+
+    int minZoom = configuredMinZoom == null ? 0 : configuredMinZoom.intValue();
+    Function<SourceFeature, Object> attributeValueProducer = attributeValueProducer(attribute);
+
+    TagCriteria attrIncludeWhen = attribute.getIncludeWhen();
+    TagCriteria attrExcludeWhen = attribute.getExcludeWhen();
+
+    Predicate<SourceFeature> attributeTest =
+      attrIncludeWhen == null ? sf -> true : attrIncludeWhen.matcher();
+    Predicate<SourceFeature> attributeExcludeTest =
+      attrExcludeWhen == null ? sf -> true : attrExcludeWhen.matcher().negate();
+
+    Double minTileCoverage = attrIncludeWhen == null ? null : attribute.getMinTileCoverSize();
+
+    Function<SourceFeature, Integer> attributeZoomProducer;
+
+    if (minTileCoverage != null && minTileCoverage > 0.0) {
+      attributeZoomProducer = attributeZoomThreshold(minTileCoverage, minZoom);
+    } else {
+      attributeZoomProducer = sf -> minZoom;
+    }
+
+    return (sf, f) -> {
+      if (attributeTest.test(sf) && attributeExcludeTest.test(sf)) {
+        f.setAttrWithMinzoom(tagKey, attributeValueProducer.apply(sf), attributeZoomProducer.apply(sf));
```

Comment:
there is a performance cost to `setAttrWithMinzoom` over just `setAttr` - we should check if there's actually a minzoom configured and if not then use `setAttr`

Reply:
Resolved in b7c4fd8